modules/javafx.controls/src/main/java/javafx/scene/control/input/BehaviorBase2.java

```diff
@@ -0,0 +1,227 @@
+import javafx.scene.input.KeyCode;
+
+/**
+ * Class provides a foundation for behaviors.
+ * <p>
+ * A concrete behavior implementation should do three things:
+ * 1. provide default behavior methods (a.k.a. functions)
+ * 2. in install() method, called from Skin.install(), map control's function tags to
+ *    behavior methods, map key bindings to function tags, and add additional event handlers,
+ *    using func(), key(), and hand() methods correspondingly.
+ *    Important: no mapping should be made in the behavior constructor, only in install().
+ * <p>
+ * The base class adds a dispose() method (called from Skin.dispose()),
+ * which undoes the mapping done in install().
+ * <p>
+ * TODO rename BehaviorBase and remove IBehavior
+ */
+public abstract class BehaviorBase2<C extends Control> {
+    private C control;
+
+    /** The constructor. */
+    public BehaviorBase2() {
+    }
+    
+    /**
+     * Returns the associated Control instance.
+     * TODO rename getControl()
+     * @return the owner
+     */
+    protected final C getNode() {
+        return control;
+    }
+
+    /**
+     * Returns the input map of the associated Control.
+     * TODO rename getInputMap()
+     * @return the input map
+     */
+    protected final InputMap2 getInputMap2() {
+        return control.getInputMap2();
+    }
+
+    /**
+     * Installs this behavior.
+     * This method must be called in Skin.install() to actually install all the default mappings.
+     * @param skin the skin
+     */
+    public void install(Skin<C> skin) {
+        Objects.nonNull(skin);
+        this.control = skin.getSkinnable();
+    }
+
+    /**
+     * Disposes of this behavior.
+     */
+    public void dispose() {
+        control.getInputMap2().unregister(this);
+    }
+    
+    /**
+     * Maps a function to the function tag.
+     * This method will not override any previous mapping added by {@link #regFunc(FunctionTag,Runnable)}.
+     *
+     * @param tag the function tag
+     * @param function the function
+     */
+    protected void regFunc(FunctionTag tag, Runnable function) {
+        getInputMap2().regFunc(this, tag, function);
+    }
+
+    /**
+     * Maps a key binding to the specified function tag.
+     * A null key binding will result in no change to this input map.
+     * This method will not override a user mapping.
+     *
+     * @param k the key binding, can be null (TODO or KB.NA)
+     * @param tag the function tag
+     */
+    protected void regKey(KeyBinding2 k, FunctionTag tag) {
+        getInputMap2().regKey(this, k, tag);
+    }
+
+    /**
+     * Maps a key binding to the specified function tag.
+     * This method will not override a user mapping added by {@link #regKey(KeyBinding2,FunctionTag)}.
+     *
+     * @param code the key code to construct a {@link KeyBinding2}
+     * @param tag the function tag
+     */
+    protected void regKey(KeyCode code, FunctionTag tag) {
+        getInputMap2().regKey(this, code, tag);
+    }
+
```

Comment:
Do you think whether it makes sense to add a variant - `regKeyAndFunc(KeyCode/KeyBinding, FunctionTag, Runnable)` for ease of use while defining default mappings?

Reply:
The earlier versions had such a method.
We could add it; I just thought that
a) the resulting code might easily break 120 character limit and
b) it might be better to have a separate block that defines the functions, since it cannot have any platform conditionals.

Reply:
(a) should not be a concern as we can break the 3 arguments over 3 lines if needed.
(b) makes sense
We can add it if it is going to make our life easier while implementing all the default Behaviors.

Reply:
added methods, thanks!